<a href="https://colab.research.google.com/github/pi-mishra/Bike-Sharing-Demand-Prediction/blob/main/Bike_Sharing_Demand_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**    - Bike Sharing Demand Prediction



##### **Project Type**    - Regression and Time series

# **Project Summary -**

# **GitHub Link -**

# **Problem Statement**

Business Context

Currently Rental bikes are introduced in many urban cities for the enhancement of mobility comfort. It is important to make the rental bike available and accessible to the public at the right time as it lessens the waiting time. Eventually, providing the city with a stable supply of rental bikes becomes a major concern. The crucial part is the prediction of bike count required at each hour for the stable supply of rental bikes.



---


Data Description:

Date = Date on which bike is on demand

Hour = Hour of day (0-23)

Temperature = Temperature of the day (C)

Humidity = Humidity measure

Windspeed = Windspeed

Visibility = Visibility measure

Dew Point Temperature = Dew Point Temperature Measure

Solar Radiation = Solar Radiation Measure

Rainfall = Rainfall in mm

Snowfall = Snowfall measure

Seasons = 1 = spring, 2 = summer, 3 = fall, 4 = winter

Holiday = Whether a holiday or not

Functional Day= Whether a functional day or not


# **1. Knowing the Data**


In [1]:
# Importing important packages

import pandas as pd
import numpy as np

In [2]:
# reading the data

df = pd.read_csv('/content/drive/MyDrive/Bike Sharing Demand Prediction/SeoulBikeData (1).csv', encoding='latin-1')

In [3]:
# top 5 rows

df.head()

Date  Rented Bike Count  Hour  Temperature(°C)  Humidity(%)  \
0  01/12/2017                254     0             -5.2           37   
1  01/12/2017                204     1             -5.5           38   
2  01/12/2017                173     2             -6.0           39   
3  01/12/2017                107     3             -6.2           40   
4  01/12/2017                 78     4             -6.0           36   

   Wind speed (m/s)  Visibility (10m)  Dew point temperature(°C)  \
0               2.2              2000                      -17.6   
1               0.8              2000                      -17.6   
2               1.0              2000                      -17.7   
3               0.9              2000                      -17.6   
4               2.3              2000                      -18.6   

   Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm) Seasons     Holiday  \
0                      0.0           0.0            0.0  Winter  No Holiday   
1                      0.0           0.0            0.0  Winter  No Holiday   
2                      0.0           0.0            0.0  Winter  No Holiday   
3                      0.0           0.0            0.0  Winter  No Holiday   
4                      0.0           0.0            0.0  Winter  No Holiday   

  Functioning Day  
0             Yes  
1             Yes  
2             Yes  
3             Yes  
4             Yes

In [4]:
# rows and column count

num_rows,num_columns = df.shape
print(f"The dataframe contains {num_rows} rows and {num_columns} columns")

The dataframe contains 8760 rows and 14 columns


In [5]:
# dataset information

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Date                       8760 non-null   object 
 1   Rented Bike Count          8760 non-null   int64  
 2   Hour                       8760 non-null   int64  
 3   Temperature(°C)            8760 non-null   float64
 4   Humidity(%)                8760 non-null   int64  
 5   Wind speed (m/s)           8760 non-null   float64
 6   Visibility (10m)           8760 non-null   int64  
 7   Dew point temperature(°C)  8760 non-null   float64
 8   Solar Radiation (MJ/m2)    8760 non-null   float64
 9   Rainfall(mm)               8760 non-null   float64
 10  Snowfall (cm)              8760 non-null   float64
 11  Seasons                    8760 non-null   object 
 12  Holiday                    8760 non-null   object 
 13  Functioning Day            8760 non-null   objec

In [6]:
# missing values

df.isnull().sum()

Date                         0
Rented Bike Count            0
Hour                         0
Temperature(°C)              0
Humidity(%)                  0
Wind speed (m/s)             0
Visibility (10m)             0
Dew point temperature(°C)    0
Solar Radiation (MJ/m2)      0
Rainfall(mm)                 0
Snowfall (cm)                0
Seasons                      0
Holiday                      0
Functioning Day              0
dtype: int64

In [7]:
# dublicate rows

df.duplicated().sum()

0

Observation-

*   The dataframe contains 8760 rows and 14 columns
*   There are no null or duplicate columns
*   Date column is not in date/time format
*   Seasons, holiday and functioning day are in object format.



# **2. Understanding Variables**

In [8]:
# column names

df.columns

Index(['Date', 'Rented Bike Count', 'Hour', 'Temperature(°C)', 'Humidity(%)',
       'Wind speed (m/s)', 'Visibility (10m)', 'Dew point temperature(°C)',
       'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 'Snowfall (cm)', 'Seasons',
       'Holiday', 'Functioning Day'],
      dtype='object')

In [9]:
# description of df

df.describe()

Rented Bike Count         Hour  Temperature(°C)  Humidity(%)  \
count        8760.000000  8760.000000      8760.000000  8760.000000   
mean          704.602055    11.500000        12.882922    58.226256   
std           644.997468     6.922582        11.944825    20.362413   
min             0.000000     0.000000       -17.800000     0.000000   
25%           191.000000     5.750000         3.500000    42.000000   
50%           504.500000    11.500000        13.700000    57.000000   
75%          1065.250000    17.250000        22.500000    74.000000   
max          3556.000000    23.000000        39.400000    98.000000   

       Wind speed (m/s)  Visibility (10m)  Dew point temperature(°C)  \
count       8760.000000       8760.000000                8760.000000   
mean           1.724909       1436.825799                   4.073813   
std            1.036300        608.298712                  13.060369   
min            0.000000         27.000000                 -30.600000   
25%            0.900000        940.000000                  -4.700000   
50%            1.500000       1698.000000                   5.100000   
75%            2.300000       2000.000000                  14.800000   
max            7.400000       2000.000000                  27.200000   

       Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm)  
count              8760.000000   8760.000000    8760.000000  
mean                  0.569111      0.148687       0.075068  
std                   0.868746      1.128193       0.436746  
min                   0.000000      0.000000       0.000000  
25%                   0.000000      0.000000       0.000000  
50%                   0.010000      0.000000       0.000000  
75%                   0.930000      0.000000       0.000000  
max                   3.520000     35.000000       8.800000

In [10]:
df['Rented Bike Count'].unique

<bound method Series.unique of 0        254
1        204
2        173
3        107
4         78
        ... 
8755    1003
8756     764
8757     694
8758     712
8759     584
Name: Rented Bike Count, Length: 8760, dtype: int64>

In [11]:
for column in df.columns:
    unique_values = df[column].unique()
    print(f"Column: {column}\nUnique Values: {unique_values}\n")

Column: Date
Unique Values: ['01/12/2017' '02/12/2017' '03/12/2017' '04/12/2017' '05/12/2017'
 '06/12/2017' '07/12/2017' '08/12/2017' '09/12/2017' '10/12/2017'
 '11/12/2017' '12/12/2017' '13/12/2017' '14/12/2017' '15/12/2017'
 '16/12/2017' '17/12/2017' '18/12/2017' '19/12/2017' '20/12/2017'
 '21/12/2017' '22/12/2017' '23/12/2017' '24/12/2017' '25/12/2017'
 '26/12/2017' '27/12/2017' '28/12/2017' '29/12/2017' '30/12/2017'
 '31/12/2017' '01/01/2018' '02/01/2018' '03/01/2018' '04/01/2018'
 '05/01/2018' '06/01/2018' '07/01/2018' '08/01/2018' '09/01/2018'
 '10/01/2018' '11/01/2018' '12/01/2018' '13/01/2018' '14/01/2018'
 '15/01/2018' '16/01/2018' '17/01/2018' '18/01/2018' '19/01/2018'
 '20/01/2018' '21/01/2018' '22/01/2018' '23/01/2018' '24/01/2018'
 '25/01/2018' '26/01/2018' '27/01/2018' '28/01/2018' '29/01/2018'
 '30/01/2018' '31/01/2018' '01/02/2018' '02/02/2018' '03/02/2018'
 '04/02/2018' '05/02/2018' '06/02/2018' '07/02/2018' '08/02/2018'
 '09/02/2018' '10/02/2018' '11/02/2018' '12/02/2

# **3. Data Wrangling**

In [14]:
# changing the date time format of column "Date"

df["Date"] = pd.to_datetime(df["Date"], format="%d/%m/%Y")

In [18]:
# copy data for wrangling

dw = df.copy()

In [27]:
# average bike count per hour

dw.groupby("Hour")["Rented Bike Count"].mean()

Hour
0      541.460274
1      426.183562
2      301.630137
3      203.331507
4      132.591781
5      139.082192
6      287.564384
7      606.005479
8     1015.701370
9      645.983562
10     527.821918
11     600.852055
12     699.441096
13     733.246575
14     758.824658
15     829.186301
16     930.621918
17    1138.509589
18    1502.926027
19    1195.147945
20    1068.964384
21    1031.449315
22     922.797260
23     671.126027
Name: Rented Bike Count, dtype: float64

In [33]:
dw.groupby(["Hour","Holiday"])["Rented Bike Count"].mean().reset_index()

Hour     Holiday  Rented Bike Count
0      0     Holiday         466.166667
1      0  No Holiday         545.365994
2      1     Holiday         393.333333
3      1  No Holiday         427.887608
4      2     Holiday         303.888889
5      2  No Holiday         301.512968
6      3     Holiday         212.333333
7      3  No Holiday         202.864553
8      4     Holiday         128.388889
9      4  No Holiday         132.809798
10     5     Holiday          89.000000
11     5  No Holiday         141.680115
12     6     Holiday         132.833333
13     6  No Holiday         295.590778
14     7     Holiday         214.277778
15     7  No Holiday         626.325648
16     8     Holiday         348.500000
17     8  No Holiday        1050.311239
18     9     Holiday         379.222222
19     9  No Holiday         659.821326
20    10     Holiday         428.555556
21    10  No Holiday         532.971182
22    11     Holiday         510.000000
23    11  No Holiday         605.564841
24    12     Holiday         621.611111
25    12  No Holiday         703.478386
26    13     Holiday         702.666667
27    13  No Holiday         734.832853
28    14     Holiday         734.000000
29    14  No Holiday         760.112392
30    15     Holiday         747.777778
31    15  No Holiday         833.409222
32    16     Holiday         787.611111
33    16  No Holiday         938.040346
34    17     Holiday         830.500000
35    17  No Holiday        1154.487032
36    18     Holiday         800.666667
37    18  No Holiday        1539.354467
38    19     Holiday         721.944444
39    19  No Holiday        1219.694524
40    20     Holiday         713.777778
41    20  No Holiday        1087.389049
42    21     Holiday         678.333333
43    21  No Holiday        1049.766571
44    22     Holiday         596.833333
45    22  No Holiday         939.706052
46    23     Holiday         451.944444
47    23  No Holiday         682.495677

In [34]:
dw.groupby("Holiday")["Rented Bike Count"].mean().reset_index()

Holiday  Rented Bike Count
0     Holiday         499.756944
1  No Holiday         715.228026

In [28]:
dw.head()

Date  Rented Bike Count  Hour  Temperature(°C)  Humidity(%)  \
0 2017-01-12                254     0             -5.2           37   
1 2017-01-12                204     1             -5.5           38   
2 2017-01-12                173     2             -6.0           39   
3 2017-01-12                107     3             -6.2           40   
4 2017-01-12                 78     4             -6.0           36   

   Wind speed (m/s)  Visibility (10m)  Dew point temperature(°C)  \
0               2.2              2000                      -17.6   
1               0.8              2000                      -17.6   
2               1.0              2000                      -17.7   
3               0.9              2000                      -17.6   
4               2.3              2000                      -18.6   

   Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm) Seasons     Holiday  \
0                      0.0           0.0            0.0  Winter  No Holiday   
1                      0.0           0.0            0.0  Winter  No Holiday   
2                      0.0           0.0            0.0  Winter  No Holiday   
3                      0.0           0.0            0.0  Winter  No Holiday   
4                      0.0           0.0            0.0  Winter  No Holiday   

  Functioning Day  
0             Yes  
1             Yes  
2             Yes  
3             Yes  
4             Yes